pkg/api/validate/quota.go

```diff
@@ -0,0 +1,93 @@
+package validate
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sort"
+
+	mgmtcompute "github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2019-03-01/compute"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/compute"
+)
+
+// checkQuota checks usage quotas vs. resources required by cluster before cluster creation
+func (dv *openShiftClusterDynamicValidator) validateQuotas(ctx context.Context, oc *api.OpenShiftCluster) error {
+	//prepare usage client
+	spp := &oc.Properties.ServicePrincipalProfile
+	r, err := azure.ParseResourceID(oc.ID)
+	if err != nil {
+		return err
+	}
+	conf := auth.NewClientCredentialsConfig(spp.ClientID, string(spp.ClientSecret), spp.TenantID)
+	authorizer, err := conf.Authorizer()
+	if err != nil {
+		return err
+	}
+	usageClient := compute.NewUsageClient(r.SubscriptionID, authorizer)
+	return dv.calculateAndCompareQuota(ctx, oc, usageClient)
+}
+
+func (dv *openShiftClusterDynamicValidator) addRequiredResources(requiredResources map[string]int, vmSize api.VMSize, count int) error {
+	requiredResources["virtualMachines"] += count
+	requiredResources["PremiumDiskCount"] += count
+	switch vmSize {
+	case api.VMSizeStandardD2sV3:
+		requiredResources["standardDSv3Family"] += count
+		requiredResources["cores"] += (count * 2)
+	case api.VMSizeStandardD4sV3:
+		requiredResources["standardDSv3Family"] += count
+		requiredResources["cores"] += (count * 4)
+	case api.VMSizeStandardD8sV3:
+		requiredResources["standardDSv3Family"] += count
+		requiredResources["cores"] += (count * 8)
+	default:
+		//will only happen if pkg/api verification allows new VMSizes
+		return fmt.Errorf("Unexpected node VMSize %s.", vmSize)
```

Comment:
internal error messages start with a lower case character and do not end with a `.` (rationale: multiple error messages are composable like `foo: bar: unimplemented vmSize %s`

Reply:
please add a comment in pkg/api to anyone who wants to add more VM types that they need to update this as well.